PERF: Use grid tree for counting grid cell inclusion

matthewturk opened the pull request:
Instead of using the grid objects, by default let's use the grid tree to count the number of cells included in a data object.

This is an independent PR, and may have test failures that can be corrected within this PR.  It's independent of #4525, but related.

matthewturk commented:
Looks like some real failures, which I can probably replicate here and work on!

matthewturk commented:
Current status is that there's a difference in the slice methods for the grid tree and the grid objects, which leads to a mismatch of roughly 1/2 of the first dimension in one of the slices.  Working on tracking it down.

matthewturk commented:
I'm working on what I hope is among the last set of issues, and it's pointing to some uninitialized memory somewhere.

matthewturk commented:
This passes locally, but answer tests may show other issues.

matthewturk commented:
To be honest, I am a little stunned that we are seeing >1k failures on `py38` but none on the unit tests.

They all follow this rough backtrace, from what I can tell:

```
  File "/home/fido/workspace/yt_py38_git/yt/data_objects/construction_data_containers.py", line 551, in __init__
    self.get_data(field)
  File "/home/fido/workspace/yt_py38_git/yt/data_objects/construction_data_containers.py", line 277, in get_data
    for chunk in parallel_objects(
  File "/home/fido/workspace/yt_py38_git/yt/utilities/parallel_tools/parallel_analysis_interface.py", line 538, in parallel_objects
    for result_id, obj in oiter:
  File "/home/fido/workspace/yt_py38_git/yt/data_objects/selection_objects/data_selection_objects.py", line 95, in chunks
    self.get_data()  # Ensure we have built ourselves
  File "/home/fido/workspace/yt_py38_git/yt/data_objects/selection_objects/data_selection_objects.py", line 140, in get_data
    self.index._identify_base_chunk(self)
  File "/home/fido/workspace/yt_py38_git/yt/geometry/grid_geometry_handler.py", line 350, in _identify_base_chunk
    dobj.size = self._count_selection(dobj, fast_index=self.grid_tree)
  File "/home/fido/shiningpanda/jobs/a7d58705/virtualenvs/d41d8cd9/lib/python3.8/functools.py", line 967, in __get__
    val = self.func(instance)
  File "/home/fido/workspace/yt_py38_git/yt/geometry/grid_geometry_handler.py", line 311, in grid_tree
    parent_ind[i] = grid.Parent.id - grid.Parent._id_offset
AttributeError: 'list' object has no attribute 'id'
```

neutrinoceros commented:
> To be honest, I am a little stunned that we are seeing >1k failures on py38 but none on the unit tests.

me too. I suspect this could be indicative of a problem with the answer test framework rather than an actual issue with your refactor, but that's just a wild (and vague) guess

matthewturk commented:
Actually, I figured it out.  It's only relevant in situations where grids can have multiple parents, but there it occurs.  I'm going to address that.

matthewturk commented:
I have to handle something else for a bit, but I just pushed a change that disabled AMRVAC using the Grid Tree and that should enable AMR types with multiple parents for a child to work.  I had to do some git stashing so it's possible I missed something, but I will return to this ASAP.

I'm going to open an issue about AMRVAC, because I think we could potentially have it load the non-leaf node AMR data as well, which would allow it to use the grid tree.

matthewturk commented:
Interestingly, passing an array as a parameter defined in a Cython argument list as `int` will throw an error, because of the conversion / sequence / scalar conversion.

matthewturk commented:
The image tests are related to #4540 and I do not believe were introduced here.

matthewturk commented:
If this passes, could I get a review?

neutrinoceros commented:
At the moment it seems that tests on Jenkins are hanging (or taking much longer than expected). I'll try to retrigger : @yt-fido test this please

matthewturk commented:
I edited some cython. This likely needs to be updated as we have with other
modules. I'll take a look.

On Tue, Aug 22, 2023, 2:08 PM Clément Robert ***@***.***>
wrote:

> At the moment it seems that tests on Jenkins are hanging (or taking much
> longer than expected). I'll try to retrigger : @yt-fido
> <https://github.com/yt-fido> test this please
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yt-project/yt/pull/4529#issuecomment-1688774678>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAVXO7YVKWGJVSKJGAMCNDXWT7TLANCNFSM6AAAAAAZO5KKIM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

matthewturk commented:
Looks like there's more to be done, according to this side-by-side comparison of recent runs.

![Screenshot from 2023-08-22 16-40-46](https://github.com/yt-project/yt/assets/89019/95f8443f-cee4-4af8-b95e-a9090f4d9bce)
![Screenshot from 2023-08-22 16-40-43](https://github.com/yt-project/yt/assets/89019/7cd420a0-1f9a-4588-a10a-632f682ff687)

Xarthisius commented:
> me too. I suspect this could be indicative of a problem with the answer test framework rather than an actual issue with your refactor, but that's just a wild (and vague) guess

Comments like that really make me want to just turn off Jenkins and be done with it. I'll note that every single failure on this PR (including most recent ones) were genuine issues that have to be addressed. As an example:

```python
import yt

uo_AM06 = {
    "length_unit": (1.0, "kpc"),
    "mass_unit": (1.0, "Msun"),
    "time_unit": (1.0, "Myr"),
}

AM06 = "AM06/AM06.out1.00400.athdf"
ds1 = yt.load(AM06, units_override=uo_AM06, magnetic_normalization="gaussian")
sp1 = ds1.sphere("c", (100.0, "kpc"))
sp1["athena_pp", "Bcc1"]
```

currently yields:

```
yt : [WARNING  ] 2023-08-23 10:35:02,366 Overriding code units: Use this option only if you know that the dataset doesn't define the units correctly or at all.
yt : [INFO     ] 2023-08-23 10:35:02,366 Overriding length_unit: 1.0 kpc.
yt : [INFO     ] 2023-08-23 10:35:02,367 Overriding time_unit: 1.0 Myr.
yt : [INFO     ] 2023-08-23 10:35:02,367 Overriding mass_unit: 1.0 Msun.
yt : [WARNING  ] 2023-08-23 10:35:02,367 Assuming 1.0 = 1.0 K
yt : [INFO     ] 2023-08-23 10:35:02,422 Parameters: current_time              = 4000.0320381013885
yt : [INFO     ] 2023-08-23 10:35:02,422 Parameters: domain_dimensions         = [128 128 128]
yt : [INFO     ] 2023-08-23 10:35:02,423 Parameters: domain_left_edge          = [-2000. -2000. -2000.]
yt : [INFO     ] 2023-08-23 10:35:02,423 Parameters: domain_right_edge         = [2000. 2000. 2000.]
yt : [INFO     ] 2023-08-23 10:35:02,423 Parameters: cosmological_simulation   = 0
Traceback (most recent call last):
  File "/home/xarth/yt/foo/bug.py", line 12, in <module>
    sp1["athena_pp", "Bcc1"]
  File "/home/xarth/codes/yt-project/yt/yt/data_objects/data_containers.py", line 235, in __getitem__
    self.get_data(f)
  File "/home/xarth/codes/yt-project/yt/yt/data_objects/selection_objects/data_selection_objects.py", line 140, in get_data
    self.index._identify_base_chunk(self)
  File "/home/xarth/codes/yt-project/yt/yt/geometry/grid_geometry_handler.py", line 361, in _identify_base_chunk
    dobj.size = self._count_selection(dobj, fast_index=self.grid_tree)
  File "/usr/lib/python3.10/functools.py", line 981, in __get__
    val = self.func(instance)
  File "/home/xarth/codes/yt-project/yt/yt/geometry/grid_geometry_handler.py", line 325, in grid_tree
    return GridTree(
  File "yt/geometry/grid_container.pyx", line 106, in yt.geometry.grid_container.GridTree.__cinit__
RuntimeError: (448, 448)
```

Maybe the fact that we have 11 green checkmarks for non-Jenkins tests is a testament to their utility.

matthewturk commented:
Still looking at the `enzo_e` errors, but the `athena_pp` errors are because athena++ only stores the leaf grids, which breaks our hierarchical search.

matthewturk commented:
Enzo-E has negative-level grids, which confused how we were computing the number of "root" grids.

matthewturk commented:
I'm still attempting to get this working; I think I'm running into some fencepost errors.  I am glad the tests are catching it quickly.

neutrinoceros commented:
@yt-fido test this please

matthewturk commented:
That's an interesting one:

```
Traceback (most recent call last):
  File "/home/fido/workspace/yt_py310_git/tests/nose_runner.py", line 9, in <module>
    from yt.config import ytcfg
  File "/home/fido/workspace/yt_py310_git/yt/__init__.py", line 13, in <module>
    from yt.data_objects.api import (
  File "/home/fido/workspace/yt_py310_git/yt/data_objects/api.py", line 1, in <module>
    from . import construction_data_containers as __cdc, selection_objects as __sdc
  File "/home/fido/workspace/yt_py310_git/yt/data_objects/construction_data_containers.py", line 17, in <module>
    from yt.data_objects.selection_objects.data_selection_objects import (
  File "/home/fido/workspace/yt_py310_git/yt/data_objects/selection_objects/__init__.py", line 1, in <module>
    from .boolean_operations import (
  File "/home/fido/workspace/yt_py310_git/yt/data_objects/selection_objects/boolean_operations.py", line 5, in <module>
    from yt.data_objects.selection_objects.data_selection_objects import (
  File "/home/fido/workspace/yt_py310_git/yt/data_objects/selection_objects/data_selection_objects.py", line 32, in <module>
    from yt.utilities.lib.marching_cubes import march_cubes_grid, march_cubes_grid_flux
ImportError: /home/fido/workspace/yt_py310_git/yt/utilities/lib/marching_cubes.cpython-310-x86_64-linux-gnu.so: undefined symbol: _Z13eval_gradientPiPdS0_S0_
```

neutrinoceros commented:
@matthewturk that's a flaky build error we've been seeing since Jenkins was migrated to Python 3.10 . IRCC the core problem is with parallel builds, but it seems arch-specific

neutrinoceros commented on yt/geometry/grid_container.pyx:
```diff
@@ -55,12 +55,23 @@ cdef class GridTree:
                   np.ndarray[np.float64_t, ndim=2] right_edge,
                   np.ndarray[np.int32_t, ndim=2] dimensions,
-                  np.ndarray[np.int64_t, ndim=1] parent_ind,
+                  list parent_ind, # Not an array anymore!
```
This comment will probably not age well. Also, I think Cython 3 supports more expressive uses of Python type annotations like `list[list[float]]`, might be worth a try
```suggestion
                  list parent_ind,
```

matthewturk replied:
I didn't know that -- I'll give that a go.

neutrinoceros commented on yt/frontends/stream/definitions.py:
```diff
@@ -47,20 +47,21 @@ def assign_particle_data(ds, pdata, bbox):
             ds.stream_handler.right_edges,
             ds.stream_handler.dimensions,
-            ds.stream_handler.parent_ids,
+            parent_ids,  # This is a list of lists
```
type checkers and modern IDEs already know that at this point :)
```suggestion
            parent_ids,
```

neutrinoceros commented on yt/geometry/_selection_routines/ray_selector.pxi:
```diff
@@ -74,6 +75,61 @@ cdef class RaySelector(SelectorObject):
+    cdef void visit_grid_cells(self, GridVisitorData *data,
+                              grid_visitor_function *func,
+                              np.uint8_t *cached_mask = NULL) noexcept nogil:
+        # AT PRESENT THIS DOES NOT WORK.
```
either this comment is now false and should be removed, or I should stop my review here and wait some more. Which is it ?

matthewturk replied:
Neither.  This method is disallowed at a higher-level, by disabling the grid visitor for ray objects, but this code should work and would be a nice breadcrumb for someone else to pick up.

neutrinoceros reviewed (commented):
I got started reviewing but stopped when I stumbled upon a comment that said the code didn't work. Waiting for further instructions :)

matthewturk commented on yt/data_objects/index_subobjects/grid_patch.py:
```diff
@@ -126,10 +127,11 @@ def _setup_dx(self):
-            if not hasattr(self.Parent, "dds"):
-                self.Parent._setup_dx()
-            self.dds = self.Parent.dds.d / self.ds.refine_by
+        parents = list(always_iterable(self.Parent, base_type=(AMRGridPatch,)))
```
It is indeed necessary, because `AMRGridPatch` exposes an iteration-like interface.  This is a long-time issue, unfortunately.

matthewturk commented:
I'm trying to let the tests go so we have a current version that includes #4525.

chrishavlin commented on yt/data_objects/tests/test_dataset_access.py:
```diff
@@ -1,15 +1,12 @@
 import numpy as np
-from nose.tools import assert_raises
```
will have to add `test_dataset_access` to the nose ignores in `nose_ignores.txt` and `tests/tests.yaml`

chrishavlin commented on yt/data_objects/tests/test_dataset_access.py:
```diff
@@ -1,15 +1,12 @@
 import numpy as np
-from nose.tools import assert_raises
+import pytest
```
will have to add `test_dataset_access` to the nose ignores in `nose_ignores.txt` and `tests/tests.yaml`

matthewturk commented:
I think I've addressed all the comments.  Unfortunately I still need to wait for the tests to pass or fail to make any more efforts to make it working :)

matthewturk commented:
Tests all pass!
